File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/ITimePeriodCollection.java
package kr.kth.commons.timeperiod;

import org.joda.time.DateTime;

/**
 * {@link ITimePeriod} 컬렉션입니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 9. 18
 */
public interface ITimePeriodCollection extends ITimePeriodContainer {

    boolean hasInsidePeriods(final ITimePeriod target);

    boolean hasOverlapPeriods(final ITimePeriod target);

    boolean hasIntersectionPeriods(final DateTime moment);

    boolean hasIntersectionPeriods(final ITimePeriod target);

    Iterable<ITimePeriod> insidePeriods(final ITimePeriod target);

    Iterable<ITimePeriod> overlapPeriods(final ITimePeriod target);

    Iterable<ITimePeriod> intersectionPeriods(final DateTime moment);

    Iterable<ITimePeriod> intersectionPeriods(final ITimePeriod target);

    Iterable<ITimePeriod> relationPeriods(final ITimePeriod target, PeriodRelation relation);

    Iterable<ITimePeriod> relationPeriods(final ITimePeriod target, PeriodRelation... relations);
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/ITimePeriodCollection.scala
package kr.kth.timeperiod

import PeriodRelation._
import org.joda.time.DateTime
import scala.annotation.varargs

/**
 * kr.kth.timeperiod.ITimePeriodCollection
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 13.
 */
trait ITimePeriodCollection extends ITimePeriodContainer {

  def hasInsidePeriods(moment: DateTime): Boolean =
    getPeriods.exists(p => Times.hasInside(p, moment))

  def hasInsidePeriods(target: ITimePeriod): Boolean =
    getPeriods.exists(p => Times.hasInside(p, target))

  def hasOverlapPeriods(target: ITimePeriod): Boolean =
    getPeriods.exists(p => Times.overlapsWith(p, target))

  def hasIntersectionPeriods(moment: DateTime): Boolean =
    hasIntersectionPeriods(TimeRange(start = moment, end = moment))

  def hasIntersectionPeriods(target: ITimePeriod): Boolean =
    getPeriods.exists(p => Times.intersectsWith(p, target))

  def insidePeriods(target: ITimePeriod): Iterable[ITimePeriod] =
    getPeriods.filter(p => Times.hasInside(p, target))

  def overlapPeriods(target: ITimePeriod): Iterable[ITimePeriod] =
    getPeriods.filter(p => Times.overlapsWith(p, target))

  def intersectionPeriods(moment: DateTime): Iterable[ITimePeriod] =
    intersectionPeriods(TimeRange(start = moment, end = moment))

  def intersectionPeriods(target: ITimePeriod): Iterable[ITimePeriod] =
    getPeriods.filter(p => Times.intersectsWith(p, target))

  def relationPeriods(target: ITimePeriod, relation: PeriodRelation): Iterable[ITimePeriod] =
    getPeriods.filter(p => Times.getRelation(p, target) == relation)

  @varargs
  def relationPeriods(target: ITimePeriod, relations: PeriodRelation*): Iterable[ITimePeriod] = {
    getPeriods.filter(p => relations.contains(Times.getRelation(p, target)))
  }
}
